cmd/cstor-pool-mgmt/controller/new-pool-controller/handler.go

```diff
@@ -0,0 +1,245 @@
+/*
+Copyright 2018 The OpenEBS Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package poolcontroller
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/golang/glog"
+	"github.com/openebs/maya/cmd/cstor-pool-mgmt/controller/common"
+
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	apis2 "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha2"
+
+	pool "github.com/openebs/maya/pkg/cstor/pool/v1alpha2"
+	corev1 "k8s.io/api/core/v1"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+)
+
+// syncHandler compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the cStorPoolUpdated resource
+// with the current status of the resource.
+func (c *CStorPoolController) syncHandler(key string, operation common.QueueOperation) error {
+	csp, err := c.getCSPObjFromKey(key)
+	if err != nil {
+		return err
+	}
+
+	return c.eventHandler(operation, csp)
+}
+
+// cStorPoolEventHandler is to handle cstor pool related events.
+func (c *CStorPoolController) eventHandler(operation common.QueueOperation, csp *apis2.CStorNPool) error {
+	switch operation {
+	case common.QOpAdd:
+		return c.addEventHandler(csp)
+	case common.QOpDestroy:
+		return c.destroyEventHandler(csp)
+	case common.QOpSync:
+		return c.syncEventHandler(csp)
+	case common.QOpModify:
+		return c.modifyEventHandler(csp)
+	}
+
+	return errors.Errorf("Invalid event {%s} for cStorPool {%s}", operation, csp.Name)
+}
+
+func (c *CStorPoolController) addEventHandler(csp *apis2.CStorNPool) error {
+	var status string
+	var err error
+	var isImported bool
+
+	uid := string(csp.GetUID())
+	importedPool := pool.ImportedCStorPools[uid]
+
+	// TODO
+	// In which scenario it is possible??
+	if importedPool != nil {
+		c.recorder.Event(importedPool,
+			corev1.EventTypeWarning,
+			string(common.AlreadyPresent),
+			string(common.MessageResourceAlreadyPresent))
+		glog.Warningf("Object is already present", string(csp.ObjectMeta.Name))
+	}
+
+	// take a lock for common
+	// TODO move return to label and unlock
+	common.SyncResources.Mux.Lock()
+
+	status, isImported = pool.IsPoolImported(csp, true /* wait if Zrepl is initializing */)
```

Comment:
why are we not wrapping csp which is the api object into our custom csp object & then handling the operations?
This is as per idiomatic maya practices.

in other words, operations will look like:
```go
import (
  csp "github.com/openebs/maya/pkg/cstor/pool/v1alpha2"
)

func (c *CStorPoolController) addEventHandler(obj *apis2.CStorNPool) error {
  // ...
  cspObj := csp.NewForAPIObject(obj)
  if cspObj.IsImported() {...}

  csp.ZFSClient().Import(cspObj)
  // ...
}
```

Reply:
We are using csp object as it as.
It is doable to wrap this object into new object and perform operation to that object, but I don't find any need to wrap this object.
Here, i have used `pool` package which operates on csp object to avoid creating new instance of object on each operation.
I agree that using wraping.. it increases the readability but using `pool.Create(csp object)` is also same. 
Still.. Once builder pattern for CSP is done, we will change this.

Reply:
@AmitKumarDas .. can you explain about ZFSClient()? code in the PR and the one you provided looks same except that argument to function is apiObject in the PR, and, it is builder object in your above code.. (or) do you think this change is argument helps in handling operations?

Reply:
I was thinking on these lines.

csp is the import path to CSP package that should have all the core modulaized business logic specific to CSP.

Caller can use builders, predicates, clients of this package.

For example:
```go
csp.New()
csp.NewBuilder()
csp.BuilderForAPIObject(api)
csp.KubeClient().Get(..)
csp.ZFSClient().Create(..)
```

Reply:
And IMO this change in approach helps is big time. A dedicated structure with receiver methods and a dedicated structure that has various typed functions based out of this dedicated structure provides more flexibility.

Reply:
We are also limited by the ability to use only a single name. So I went for our custom structure.

We might also rethink and try to define multiple names like `IsImported` and `IsImportedForAPI`.